clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
   D->addAttr(::new (S.Context) PreferredTypeAttr(S.Context, AL, ParmTSI));
 }
 
+// This function is called only if function call is not inside template body.
+// TODO: Add call for function calls inside template body.
+// Emit warnings if parent function misses format attributes.
+void Sema::DiagnoseMissingFormatAttributes(Stmt *Body,
+                                           const FunctionDecl *FDecl) {
+  assert(FDecl);
+
+  // If there are no function body, exit.
+  if (!Body)
+    return;
+
+  // Get missing format attributes
+  std::vector<FormatAttr *> MissingFormatAttributes =
+      GetMissingFormatAttributes(Body, FDecl);
+  if (MissingFormatAttributes.empty())
+    return;
+
+  // Check if there are more than one format type found. In that case do not
+  // emit diagnostic.
+  const clang::IdentifierInfo *AttrType = MissingFormatAttributes[0]->getType();
+  if (llvm::any_of(MissingFormatAttributes, [&](const FormatAttr *Attr) {
+        return AttrType != Attr->getType();
+      }))
+    return;
+
+  for (const FormatAttr *FA : MissingFormatAttributes) {
+    // If format index and first-to-check argument index are negative, it means
+    // that this attribute is only saved for multiple format types checking.
+    if (FA->getFormatIdx() < 0 || FA->getFirstArg() < 0)
+      continue;
+
+    // Emit diagnostic
+    SourceLocation Loc = FDecl->getLocation();
+    Diag(Loc, diag::warn_missing_format_attribute)
+        << FA->getType() << FDecl
+        << FixItHint::CreateInsertion(Loc,
+                                      (llvm::Twine("__attribute__((format(") +
+                                       FA->getType()->getName() + ", " +
+                                       llvm::Twine(FA->getFormatIdx()) + ", " +
+                                       llvm::Twine(FA->getFirstArg()) + ")))")
+                                          .str());
+  }
+}
+
+// Returns vector of format attributes. There are no two attributes with same
+// arguments in returning vector. There can be attributes that effectivelly only
+// store information about format type.
+std::vector<FormatAttr *>
+Sema::GetMissingFormatAttributes(Stmt *Body, const FunctionDecl *FDecl) {
+  unsigned int FunctionFormatArgumentIndexOffset =
+      checkIfMethodHasImplicitObjectParameter(FDecl) ? 2 : 1;
+
+  std::vector<FormatAttr *> MissingAttributes;
+
+  // Iterate over body statements.
+  for (auto *Child : Body->children()) {
+    // If child statement is compound statement, recursively get missing
+    // attributes.
+    if (dyn_cast_or_null<CompoundStmt>(Child)) {
```

Comment:
The idea of this part of code is to enable diagnosing missing format attributes if there are branches around call, ie. call is not direct child of caller function. This part of code is aimed to handle examples like this:

```
void f(char *out, va_list args)
{
    {
        vprintf(out, args);
    }
}
```

Without this check, missing format attribute in example above would not be diagnosed. This part of code enables recursively checking children statements.

If statement is not compound, next check is if it is a function call.

Reply:
When you write about "branches around call", do you mean braces? However, I meant branches:
```c++
void f(char *out, va_list args)
{
    if (tracing_enabled) {
        vprintf(out, args);
    }
}
```
This is not a compound statement, so you'd not be looking into that.

My proposal would be to check on encountering an annotated `CallExpr` instead of traversing the function body. This is also better for performance.

Reply:
Yes, I meant braces instead of branches. This snippet does not produce warning, so I'm working to fix it.

Reply:
Changed logic to detect missing format attributes on an annotated function calls. Here are differences: [link](https://github.com/llvm/llvm-project/compare/97b70712c658bae20657dd78253d8de10319bf2d..68fb91662185dc164574f0058e660310c05e08a4).